### PR TITLE
[REVIEW] Fix `replace` to handle null values correctly

### PR DESCRIPTION
**Pull request by galipremsagar**
Fixes: #7293 

This PR introduces fixes to `replace` and it's internal APIs which error/fault when there is a `None`/`<NA>` value on either `lhs` or `rhs` and certain combinations of each. Introduced tests where there were failures previously that are covered by new changes in this PR.

### CODE REVIEW

**Comment by galipremsagar** [comment]
rerun tests

File: python/cudf/cudf/core/column/categorical.py [hunk 1]
```diff
@@ -1065,20 +1065,52 @@ def find_and_replace(
...
+            df = df.dropna(subset=["old"])
+            to_replace_col = df["old"]._column
+            replacement_col = df["new"]._column
+        else:
+            replaced = self
+        if df._data["new"].null_count > 0:
+            drop_values = df._data["old"][df._data["new"].isna()]
+            cur_categories = replaced.categories
+            new_categories = cur_categories[
+                ~cudf.Series(cur_categories.isin(drop_values))
```
Comment (isVoid) [inline python/cudf/cudf/core/column/categorical.py:1102]: I think we can reduce overhead here using `Series._from_data`.

**Review by isVoid — COMMENTED** [review]
Just small changes.

File: python/cudf/cudf/core/column/categorical.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/column/categorical.py:1102]: `cur_categories.isin(drop_values)` result is a `Column` here

File: python/cudf/cudf/core/column/string.py [hunk 2]
```diff
@@ -5277,8 +5277,16 @@ def find_and_replace(
...
-        return libcudf.replace.replace(self, to_replace_col, replacement_col)
+        df = cudf.DataFrame({"old": to_replace_col, "new": replacement_col})
+        df = df.drop_duplicates(subset=["old"], keep="last", ignore_index=True)
+        if df._data["old"].null_count == 1:
+            res = self.fillna(df._data["new"][df._data["old"].isna()][0])
+            df = df.dropna(subset=["old"])
+        else:
+            res = self
+        return libcudf.replace.replace(
+            res, df["old"]._column, df["new"]._column
```
[the remark this answers is not in the document]
Comment (galipremsagar) [reply python/cudf/cudf/core/column/string.py]: Done

File: python/cudf/cudf/core/column/categorical.py (hunk 1 above)
Comment (isVoid) [reply python/cudf/cudf/core/column/categorical.py:1102]: Oh I was more referring to `_from_data(ColumnAccessor({None: column})`. But I guess we can delay this until https://github.com/rapidsai/cudf/pull/8687 gives us a better syntax of `_from_data({None: column})`

**Review by isVoid — APPROVED** [review]

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**